### PR TITLE
DAOS-9461 engine: return failure if cores could not satisfy XS number

### PR DESCRIPTION
**Pull request by wangshilong**
Without DAOS_TARGET_OVERSUBSCRIBE, daos engine should exit as
failure if number of cores could not meet XS requirement.

Signed-off-by: Wang Shilong <shilong.wang@intel.com>

### CODE REVIEW

File: src/engine/init.c [hunk 1]
```diff
@@ -283,9 +283,12 @@ dss_tgt_nr_get(unsigned int ncores, unsigned int nr, bool oversubscribe)
 		goto retry;
 	}
 
-	if (tgt_nr != nr)
+	if (tgt_nr != nr) {
 		D_PRINT("%d target XS(xstream) requested (#cores %d); "
 			"use (%d) target XS\n", nr, ncores, tgt_nr);
+		if (!oversubscribe)
+			return -DER_INVAL;
```
Comment (NiuYawei) [inline src/engine/init.c:290]: The console message above and the comments of 'oversubscribe' in this function needs be revised.

File: src/engine/init.c [hunk 2]
```diff
@@ -317,9 +320,15 @@ dss_topo_init()
 	/* fall back to the legacy core allocation algorithm */
 	if (dss_numa_node == -1 || numa_node_nr <= 0) {
 		D_PRINT("Using legacy core allocation algorithm\n");
-		dss_tgt_nr = dss_tgt_nr_get(dss_core_nr, nr_threads,
-					    tgt_oversub);
-
+		rc = dss_tgt_nr_get(dss_core_nr, nr_threads,
+				    tgt_oversub);
+		if (rc < 0) {
```
Comment (NiuYawei) [inline src/engine/init.c:325]: The NUMA aware case (see the bottom of this function) is missed.

**Review by NiuYawei — CHANGES_REQUESTED** [review]

File: src/engine/init.c (hunk 1 above)
Comment (NiuYawei) [reply src/engine/init.c:288]: the comments in this function should be revised as well.

**Review by NiuYawei — APPROVED** [review]

**Review by liuxuezhao — APPROVED** [review]

File: src/engine/init.c [hunk 3]
```diff
@@ -283,9 +282,12 @@ dss_tgt_nr_get(unsigned int ncores, unsigned int nr, bool oversubscribe)
 		goto retry;
 	}
 
-	if (tgt_nr != nr)
+	if (tgt_nr != nr) {
+		if (!oversubscribe)
```
Comment (NiuYawei) [inline src/engine/init.c:286]: This check is redundant, actually it should be "assert(!oversubscribe)", right?

**Review by NiuYawei — APPROVED** [review]

File: src/engine/init.c (hunk 3 above)
Comment (wangshilong) [reply src/engine/init.c:286]: No, this is needed to return errors if oversubscribe is not enabled.

**Review by liuxuezhao — APPROVED** [review]

**Review by daltonbohning — COMMENTED** [review]
ftest changes look good. Is there a unit / functional test to verify the error is returned as expected?

**Comment by wangshilong** [comment]
> ftest changes look good. Is there a unit / functional test to verify the error is returned as expected?

@daltonbohning Do you have idea if there existed similar test case for this where i could copy/learn?

**Comment by daltonbohning** [comment]
> > ftest changes look good. Is there a unit / functional test to verify the error is returned as expected?
> 
> @daltonbohning Do you have idea if there existed similar test case for this where i could copy/learn?

I think this test case/yaml could have a variant for this:
https://github.com/daos-stack/daos/blob/master/src/tests/ftest/server/daos_server_config.py
https://github.com/daos-stack/daos/blob/master/src/tests/ftest/server/daos_server_config.yaml

**Pull request closed without merging**